core/trino-main/src/main/java/io/trino/type/DecimalOperators.java

```diff
@@ -106,67 +101,48 @@ public static long addShortShortShort(long a, long b, long aRescale, long bResca
     }
 
     @UsedByGeneratedCode
-    public static Slice addShortShortLong(long a, long b, int rescale, boolean left)
+    public static Int128 addShortShortLong(long a, long b, int rescale, boolean left)
     {
-        return internalAddLongLongLong(unscaledDecimal(a), unscaledDecimal(b), rescale, left);
+        return internalAddLongLongLong(a >> 63, a, b >> 63, b, rescale, left);
     }
 
     @UsedByGeneratedCode
-    public static Slice addLongLongLong(Slice a, Slice b, int rescale, boolean left)
+    public static Int128 addLongLongLong(Int128 a, Int128 b, int rescale, boolean left)
     {
-        return internalAddLongLongLong(a, b, rescale, left);
+        return internalAddLongLongLong(a.getHigh(), a.getLow(), b.getHigh(), b.getLow(), rescale, left);
     }
 
     @UsedByGeneratedCode
-    public static Slice addShortLongLong(long a, Slice b, int rescale, boolean left)
+    public static Int128 addShortLongLong(long a, Int128 b, int rescale, boolean left)
     {
         return addLongShortLong(b, a, rescale, !left);
     }
 
     @UsedByGeneratedCode
-    public static Slice addLongShortLong(Slice a, long b, int rescale, boolean rescaleLeft)
+    public static Int128 addLongShortLong(Int128 a, long b, int rescale, boolean rescaleLeft)
     {
-        try {
-            Slice left;
-            Slice right;
-
-            if (rescaleLeft) {
-                left = unscaledDecimal();
-                rescale(a, rescale, left);
-                right = unscaledDecimal(b);
-            }
-            else {
-                left = rescale(b, rescale);
-                right = a;
-            }
-
-            add(left, right, left);
-            throwIfOverflows(left);
-            return left;
-        }
-        catch (ArithmeticException e) {
-            throw new TrinoException(NUMERIC_VALUE_OUT_OF_RANGE, "Decimal overflow", e);
-        }
+        return internalAddLongLongLong(a.getHigh(), a.getLow(), b >> 63, b, rescale, rescaleLeft);
     }
 
-    private static Slice internalAddLongLongLong(Slice a, Slice b, int rescale, boolean rescaleLeft)
+    private static Int128 internalAddLongLongLong(long leftHigh, long leftLow, long rightHigh, long rightLow, int rescale, boolean rescaleLeft)
     {
+        // TODO: specialize implementation for rescale == 0 vs rescale > 0, and rescale left vs rescale right to avoid branch pollution
         try {
-            Slice left = unscaledDecimal();
-            Slice right;
-
+            long[] result = new long[2];
```

Comment:
We need arrays because some of the decoders (ORC, I believe) decode directly into an array of longs. I didn't want to introduce two variants (one with array, one with MutableInt128) quite yet.